File: java/magical3.java
import java.util.Scanner;

public class magical3 {
	static boolean isPrime(int number) {
		for (int i = 2; i <= Math.ceil(Math.sqrt(number)); i++)
			if (number % i == 0)
				return false;
		return true;
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int input;
		while ((input = scan.nextInt()) != 0) {
			int _base = -1;

			input -= 3;
			if (input == 0)
				_base = 4;
			else if (input > 3) {
				boolean isPrime = true;
				for (int i = 2; i <= Math.ceil(Math.sqrt(input)); i++) {
					if (input % i == 0) {
						isPrime = false;
						if (i >= 4) {
							_base = i;
							break;
						} else {
							int factor = input / i;
							if (isPrime(factor) && factor >= 4) {
								_base = factor;
								break;
							}
							break;
						}
					}
				}
				if (isPrime) {
					_base = input;
				} else {
					if (_base == -1) {
						_base = 4;
						while (input >= _base) {
							if (input % _base == 0)
								break;
							_base++;
						}
						if (input < _base)
							_base = -1;
					}
				}
			}

			if (_base == -1)
				System.out.println("No such base");
			else
				System.out.println(_base);

		}
		scan.close();
	}
}

File: ts/src/magical3.ts
import readline = require('readline')

function isPrimeNumber (num: number): boolean {
  for (let i = 2; i <= Math.ceil(Math.sqrt(num)); i++) {
    if (num % i === 0) {
      return false
    }
  }
  return true
}

function getBase (num: number): number {
  num -= 3

  if (num === 0) {
    return 4
  }

  if (num <= 3) {
    return -1
  }

  let isPrime = true
  for (let i = 2; i <= Math.ceil(Math.sqrt(num)); i++) {
    if (num % i === 0) {
      isPrime = false
      if (i >= 4) {
        return i
      }
      const factor = num / i
      if (factor >= 4 && isPrimeNumber(factor)) {
        return factor
      }
      break
    }
  }

  if (isPrime) {
    return num
  }

  for (let i = 4; i <= num; i++) {
    if (num % i === 0) {
      return i
    }
  }

  return -1
}

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (line === '0') {
      process.exit(0)
    }
    const num = parseInt(line)
    const base = getBase(num)
    console.log(base === -1 ? 'No such base' : base)
  })
  .on('close', () => {
    process.exit(0)
  })
